beacon-chain/types/block.go

```diff
@@ -128,3 +140,104 @@ func (b *Block) Attestations() []*pb.AttestationRecord {
 func (b *Block) Timestamp() (time.Time, error) {
 	return ptypes.Timestamp(b.data.Timestamp)
 }
+
+// isSlotValid compares the slot to the system clock to determine if the block is valid.
+func (b *Block) isSlotValid() bool {
+	slotDuration := time.Duration(b.SlotNumber()*params.SlotDuration) * time.Second
+	validTimeThreshold := genesisTime.Add(slotDuration)
+
+	return clock.Now().After(validTimeThreshold)
+}
+
+// IsValid is called to decide if an incoming p2p block can be processed into the chain's block trie,
+// it checks time stamp, beacon chain parent block hash. It also checks pow chain reference hash if it's a validator.
+func (b *Block) IsValid(aState *ActiveState, cState *CrystallizedState) bool {
+	_, err := b.Hash()
+	if err != nil {
+		log.Debugf("Could not hash incoming block: %v", err)
+		return false
+	}
+
+	if b.SlotNumber() == 0 {
+		log.Debugf("Can not process a genesis block")
+		return false
+	}
+
+	if !b.isSlotValid() {
+		log.Debugf("slot of block is too high: %d", b.SlotNumber())
+		return false
+	}
+
+	for index, attestation := range b.Attestations() {
+		if !b.isAttestationValid(index, aState, cState) {
+			log.Debugf("attestation invalid: %v", attestation)
+			return false
+		}
+	}
+
+	return true
+}
+
+// isAttestationValid validates an attestation in a block
+func (b *Block) isAttestationValid(attestationIndex int, aState *ActiveState, cState *CrystallizedState) bool {
+	// Validate attestation's slot number has is within range of incoming block number.
+	slotNumber := int(b.SlotNumber())
+	attestation := b.Attestations()[attestationIndex]
+	if int(attestation.Slot) > slotNumber {
+		log.Debugf("attestation slot number can't be higher than block slot number. Found: %d, Needed lower than: %d",
+			attestation.Slot,
+			slotNumber)
+		return false
+	}
+	if int(attestation.Slot) < slotNumber-params.CycleLength {
+		log.Debugf("attestation slot number can't be lower than block slot number by one CycleLength. Found: %v, Needed greater than: %v",
+			attestation.Slot,
+			slotNumber-params.CycleLength)
+		return false
+	}
+
+	if attestation.JustifiedSlot > cState.LastJustifiedSlot() {
+		log.Debugf("attestation's last justified slot has to match crystallied state's last justified slot. Found: %d. Want: %d",
+			attestation.JustifiedSlot,
+			cState.LastJustifiedSlot())
+		return false
+	}
+
+	// TODO: Validate last justified block hash matches in the crystallizedState.
+
+	// Get all the block hashes up to cycle length.
+	parentHashes := aState.getSignedParentHashes(b, attestation)
+	attesterIndices, err := cState.getAttesterIndices(attestation)
+	if err != nil {
+		log.Debugf("unable to get validator committee: %v", attesterIndices)
+		return false
+	}
+
+	// Verify attester bitfields matches crystallized state's prev computed bitfield.
+	if !casper.AreAttesterBitfieldsValid(attestation, attesterIndices) {
+		return false
+	}
+
+	// TODO: Generate validators aggregated pub key.
```

Comment:
Is there a github issue created for this TODO? Please reference the issue number here

Reply:
@terenc3t , are there issues for these?